### PR TITLE
FilePathLayout - Reduce memory-allocation for cleanup of filename

### PR DESCRIPTION
**Pull request by snakefoot**
Reuse clean-filename when raw-filename is unchanged.

---
This change is [Reviewable](https://reviewable.io/reviews/nlog/nlog/1745)

### CODE REVIEW

File: src/NLog/Internal/FilePathLayout.cs [hunk 1]
```diff
@@ -77,6 +77,9 @@ internal class FilePathLayout : IRenderable
 
         private bool _cleanupInvalidChars;
 
+        private string _cachedPrevRawFileName;
+        private string _cachedPrevCleanFileName;
```
Comment (304NotModified) [inline src/NLog/Internal/FilePathLayout.cs]: We should document (xmldoc) that `_cachedPrevCleanFileName` is a pair with `_cachedPrevRawFileName`, right?

File: src/NLog/Internal/FilePathLayout.cs [hunk 2]
```diff
@@ -138,56 +141,78 @@ public Layout GetLayout()
...
+            var rawFileName = GetRenderedFileName(logEvent);
+            if (string.IsNullOrEmpty(rawFileName))
+            {
+                return rawFileName;
+            }
+
+            if ((!_cleanupInvalidChars || cleanedFixedResult != null) && _filePathKind == FilePathKind.Absolute)
+                return rawFileName; // Skip clean filename string-allocation
+
+            if (string.CompareOrdinal(_cachedPrevRawFileName, rawFileName) == 0 && _cachedPrevCleanFileName != null)
```
Comment (304NotModified) [inline src/NLog/Internal/FilePathLayout.cs]: In case of performance, the check `_cachedPrevCleanFileName != null` should be on the left?

Comment (snakefoot) [inline src/NLog/Internal/FilePathLayout.cs]: _cachedPrevCleanFileName will never be null, and when it is then _cachedPrevRawFileName is also null. I just have a secret agenda when it comes to the PR with PoolSetup.

File: src/NLog/Internal/FilePathLayout.cs (hunk 1 above)
Comment (snakefoot) [inline src/NLog/Internal/FilePathLayout.cs]: Yes they are working together. Will update the xml-doc

**Comment by 304NotModified** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1745)*

**Comment by 304NotModified** [comment]
Thanks again!

**Pull request merged**